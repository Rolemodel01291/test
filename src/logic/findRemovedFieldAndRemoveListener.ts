import removeAllEventListeners from './removeAllEventListeners';
import isRadioInput from '../utils/isRadioInput';
import { Field, FieldsObject, DataType } from '../types';

export default function findRemovedFieldAndRemoveListener<Data extends DataType>(
  fields: FieldsObject<Data>,
  validateWithStateUpdate: Function,
  { ref, mutationWatcher, options }: Field,
  forceDelete: boolean = false,
): FieldsObject<Data> | undefined {
  if (!ref || !ref.type) return;

  const { name, type } = ref;
  if (isRadioInput(type) && options) {
    options.forEach(
      ({ ref }, index): void => {
        if (ref instanceof HTMLElement && !document.body.contains(ref) && options && options[index]) {
          removeAllEventListeners(options[index], validateWithStateUpdate);
          (options[index].mutationWatcher || { disconnect: (): void => {} }).disconnect();
          options.splice(index, 1);
        }
      },
    );

    if (Array.isArray(options) && !options.length) {
      delete fields[name];
      return fields;
    }
  } else if ((ref instanceof HTMLElement && !document.body.contains(ref)) || forceDelete) {
    removeAllEventListeners(ref, validateWithStateUpdate);
    if (mutationWatcher) mutationWatcher.disconnect();
    delete fields[name];
    return fields;
  }

  return fields;
}
