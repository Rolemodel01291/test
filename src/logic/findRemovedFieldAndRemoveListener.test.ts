import findRemovedFieldAndRemoveListener from './findRemovedFieldAndRemoveListener';

jest.mock('./removeAllEventListeners');

describe('findMissDomAndClean', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should return default fields value if nothing matches', () => {
    document.body.contains = jest.fn(() => true);
    const fields = {
      test: 'test',
    };
    expect(
      // @ts-ignore
      findRemovedFieldAndRemoveListener(fields, () => {}, {
        ref: { name: 'bill', type: 'radio' },
      }),
    ).toEqual(fields);
  });

  it('should remove options completely if option found and no option left', () => {
    document.body.contains = jest.fn(() => false);

    const ref = document.createElement('input');
    ref.setAttribute('name', 'test');
    ref.setAttribute('type', 'radio');

    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        ref: {},
        options: [
          {
            ref: 'test',
            mutationWatcher: {
              disconnect,
            },
          },
        ],
      },
    };

    expect(
      // @ts-ignore
      findRemovedFieldAndRemoveListener(fields, () => {}, {
        ref: { name: 'test', type: 'radio' },
        options: [
          {
            mutationWatcher: {
              disconnect,
            },
            ref,
          },
        ],
      }),
    ).toMatchSnapshot();
  });

  it('should remove none radio field when found', () => {
    const ref = document.createElement('input');
    ref.setAttribute('name', 'test');
    ref.setAttribute('type', 'radio');
    document.body.contains = jest.fn(() => false);
    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        ref: {},
        mutationWatcher: {
          disconnect,
        },
      },
      test1: {
        name: 'test',
        ref: {},
      },
    };

    expect(
      findRemovedFieldAndRemoveListener(fields, () => {}, {
        ref,
        mutationWatcher: {
          disconnect,
        },
      }),
    ).toMatchSnapshot();
  });

  it('should return undefined when empty ref', () => {
    const fields = {
      test: 'test',
    };
    expect(
      // @ts-ignore
      findRemovedFieldAndRemoveListener(fields, () => {}, {}),
    ).toEqual(undefined);
  });

  it('should work for radio type input', () => {
    const ref = document.createElement('input');
    ref.setAttribute('name', 'test');
    ref.setAttribute('type', 'radio');
    document.body.contains = jest.fn(() => false);
    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        ref: {},
        mutationWatcher: {
          disconnect,
        },
      },
      test1: {
        name: 'test',
        ref: {
          type: 'radio',
        },
      },
    };

    expect(
      findRemovedFieldAndRemoveListener(fields, () => {}, {
        ref: { name: 'test', type: 'radio' },
        options: [{ ref }],
        mutationWatcher: {
          disconnect,
        },
      }),
    ).toMatchSnapshot();
  });

  it('should not remove event listner when type is not Element', () => {
    document.body.contains = jest.fn(() => false);

    const disconnect = jest.fn();
    const fields = {
      test: {
        name: 'test',
        ref: {},
        options: [
          {
            ref: 'test',
            mutationWatcher: {
              disconnect,
            },
          },
        ],
      },
    };

    expect(
      // @ts-ignore
      findRemovedFieldAndRemoveListener(fields, () => {}, {
        ref: { name: 'test', type: 'text' },
        options: [
          {
            mutationWatcher: {
              disconnect,
            },
            ref: {},
          },
        ],
      }),
    ).toMatchSnapshot();

    expect(
      // @ts-ignore
      findRemovedFieldAndRemoveListener(fields, () => {}, {
        ref: { name: 'test', type: 'text' },
      }),
    ).toMatchSnapshot();
  });
});
