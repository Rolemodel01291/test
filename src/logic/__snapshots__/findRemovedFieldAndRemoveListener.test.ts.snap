// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findMissDomAndClean should not remove event listner when type is not Element 1`] = `
Object {
  "test": Object {
    "name": "test",
    "options": Array [
      Object {
        "mutationWatcher": Object {
          "disconnect": [MockFunction],
        },
        "ref": "test",
      },
    ],
    "ref": Object {},
  },
}
`;

exports[`findMissDomAndClean should not remove event listner when type is not Element 2`] = `
Object {
  "test": Object {
    "name": "test",
    "options": Array [
      Object {
        "mutationWatcher": Object {
          "disconnect": [MockFunction],
        },
        "ref": "test",
      },
    ],
    "ref": Object {},
  },
}
`;

exports[`findMissDomAndClean should remove none radio field when found 1`] = `
Object {
  "test1": Object {
    "name": "test",
    "ref": Object {},
  },
}
`;

exports[`findMissDomAndClean should remove options completely if option found and no option left 1`] = `Object {}`;

exports[`findMissDomAndClean should work for radio type input 1`] = `
Object {
  "test1": Object {
    "name": "test",
    "ref": Object {
      "type": "radio",
    },
  },
}
`;
